### PR TITLE
Register filled flask as a Slimefun Item

### PR DESCRIPTION
**Pull request by qwertyuioplkjhgfd**
## Description

Currently, ``FILLED_FLASK_OF_KNOWLEDGE`` is the only Slimefun Item that has a ``SlimefunItemStack`` but is never registered as a ``SlimefunItem`` in ``SlimefunItemSetup``. This meant that calls to ``SlimefunItem.getById("FILLED_FLASK_OF_KNOWLEDGE")`` would return ``null``. This pull request registers a SlimefunItem for the ItemStack with minimal behavioral changes.

## Proposed changes

- Create ``FilledKnowledgeFlask``, which is a ``SlimefunItem`` that implements ``NotConfigurable`` and is hidden in the Slimefun guide.
- Create a new ``RecipeType``: ``RecipeType.INTERACT``, which signifies that the item is obtained by interacting with the ingredient.
- Create and register a ``FilledKnowledgeFlask`` instance in ``SlimefunItemSetup``.

The overall effects of these changes are:
- When clicking a filled flask of knowledge in the Slimefun guide, it will now display the method of obtaining a filled flask, which is interacting with an empty flask of knowledge. 
- ``FILLED_FLASK_OF_KNOWLEDGE`` will be added to the ``SlimefunRegistry``.

## Related Issues (if applicable)

``SlimefunItem.getById("FILLED_FLASK_OF_KNOWLEDGE")`` no longer returns ``null``.

## Checklist

- [x] I have fully tested the proposed changes and promise that they will not break everything into chaos.
- [x] I have also tested the proposed changes in combination with various popular addons and can confirm my changes do not break them.
- [x] I followed the existing code standards and didn't mess up the formatting.
- [ ] I did my best to add documentation to any public classes or methods I added.
- [ ] I have added `Nonnull` and `Nullable` annotations to my methods to indicate their behaviour for null values
- [ ] I added sufficient Unit Tests to cover my code.

### CODE REVIEW

File: src/main/java/io/github/thebusybiscuit/slimefun4/implementation/items/HiddenItem.java
Comment (svr333) [inline src/main/java/io/github/thebusybiscuit/slimefun4/implementation/items/HiddenItem.java:16]:
```suggestion
 * The {@link HiddenItem} is a {@link NotConfigurable} {@link SlimefunItem}
 * that is hidden from the Slimefun guide.
```

File: src/main/java/io/github/thebusybiscuit/slimefun4/implementation/items/HiddenItem.java
Comment (svr333) [inline src/main/java/io/github/thebusybiscuit/slimefun4/implementation/items/HiddenItem.java:26]:
```suggestion
}
```

**Review by svr333 — CHANGES_REQUESTED** [review]

**Review by poma123 — CHANGES_REQUESTED** [review]

File: src/main/java/io/github/thebusybiscuit/slimefun4/implementation/items/HiddenItem.java
Comment (svr333) [inline src/main/java/io/github/thebusybiscuit/slimefun4/implementation/items/HiddenItem.java:22]: couldnt we default the recipe to an empty `ItemStack[]` (or 9 nulls) meaning the ctor would get this declaration:
`public HiddenItem(ItemGroup itemGroup, SlimefunItemStack item, RecipeType recipeType)`
since this will always be hidden?

That way we dont have to initialize them in `SlimefunItemSetup` with a lot of nulls

Comment (qwertyuioplkjhgfd) [reply src/main/java/io/github/thebusybiscuit/slimefun4/implementation/items/HiddenItem.java:22]: well right now it has the empty flask as an ingredient, i guess it would be better to have it just be nulls, would that be better?

**Comment by qwertyuioplkjhgfd** [comment]
how do you squash

**Comment by svr333** [comment]
squashing is done right before we merge into another branch, so one of the maintainers has to do it

File: src/main/java/io/github/thebusybiscuit/slimefun4/implementation/items/HiddenItem.java
Comment (svr333) [reply src/main/java/io/github/thebusybiscuit/slimefun4/implementation/items/HiddenItem.java:25]: well the thing is the recipe doesnt matter right?
so we could make the constructor of HiddenItem something like:

```java
@ParametersAreNonnullByDefault
    public HiddenItem(ItemGroup itemGroup, SlimefunItemStack item, RecipeType recipeType) {
        super(itemGroup, item, recipeType, new ItemStack[] {});
        this.setHidden(true);
```

this means that if we then create an item somewhere else, we dont even have to initalize a recipe, because its not used anyway

Comment (poma123) [reply src/main/java/io/github/thebusybiscuit/slimefun4/implementation/items/HiddenItem.java:25]: > the thing is the recipe doesnt matter right?

I don't really agree. It does matter as we have a new RecipeType that shows players what item should be right-clicked with to obtain the result item (in this case the filled knowledge flask).
This only makes sense if we do show the recipe item (in this case the empty flask)

Comment (svr333) [reply src/main/java/io/github/thebusybiscuit/slimefun4/implementation/items/HiddenItem.java:25]: fair enough, but in that case should we just have a singular ItemStack then, instead of an array?

Comment (svr333) [reply src/main/java/io/github/thebusybiscuit/slimefun4/implementation/items/HiddenItem.java:25]:
```java
@ParametersAreNonnullByDefault
    public HiddenItem(ItemGroup itemGroup, SlimefunItemStack item, RecipeType recipeType, ItemStack recipe) {
        super(itemGroup, item, recipeType, new ItemStack[] { recipe, null, null, null, null, null, null, null, null });
        this.setHidden(true);
```

Comment (poma123) [reply src/main/java/io/github/thebusybiscuit/slimefun4/implementation/items/HiddenItem.java:25]: What about `recipeItem` instead?

Comment (qwertyuioplkjhgfd) [reply src/main/java/io/github/thebusybiscuit/slimefun4/implementation/items/HiddenItem.java:25]: That wouldn't work actually. Hidden items are not necessarily of RecipeType INTERACT. This would limit hidden items to only be able to have a single ingredient, which I'm not sure is the intention.

**Comment by qwertyuioplkjhgfd** [comment]
bump; not sure if the requested changes are intended

File: src/main/java/io/github/thebusybiscuit/slimefun4/implementation/items/HiddenItem.java
Comment (poma123) [reply src/main/java/io/github/thebusybiscuit/slimefun4/implementation/items/HiddenItem.java:25]: Ah, good point. We should then leave it as it was before this, I don't think that the 8 nulls would be a huge problem anyway. We have a bunch of them in that file already, I think it's okay.

**Review by svr333 — APPROVED** [review]
good for me then

**Review by poma123 — CHANGES_REQUESTED** [review]
> When clicking a filled flask of knowledge in the Slimefun guide, it will now display the method of obtaining a filled flask, which is interacting with an empty flask of knowledge.

I mean if it is hidden from the guide, how the heck would it display IN the guide how to obtain it if it is hidden already..? It's clearly not in the guide since it is a HiddenItem... This doesn't make any sense to me.

**Comment by svr333** [comment]
> I mean if it is hidden from the guide, how the heck would it display IN the guide how to obtain it if it is hidden already..? It's clearly not in the guide since it is a HiddenItem... This doesn't make any sense to me.

well yeah thats the first remark I made as well, you know if the item is hidden the recipe is never shown

**Comment by qwertyuioplkjhgfd** [comment]
> > When clicking a filled flask of knowledge in the Slimefun guide, it will now display the method of obtaining a filled flask, which is interacting with an empty flask of knowledge.
> 
> I mean if it is hidden from the guide, how the heck would it display IN the guide how to obtain it if it is hidden already..? It's clearly not in the guide since it is a HiddenItem... This doesn't make any sense to me.

it will show the recipe if you click on a filled flask in a recipe of another item, such as the talisman of the wise.

**Review by poma123 — APPROVED** [review]
Okay then.

**Review by svr333 — APPROVED** [review]

**Review by TheBusyBiscuit — APPROVED** [review]
LGTM

**Pull request merged**
